### ISSUE TITLE
File Browser (Alt+U) always takes few seconds to start 

### ISSUE THREAD

**Issue by danielHalan** [sergey-visual-studio/dpack#56]
After the switch to DPack Rx noticed a performance degradation when opening up the File Browser. 
Previously (pre-Rx) this was near to instant, but now it will take 1-4sec each time. 

One note is that the Solution in question is quite large with ~190 projects. 
Is it that file structure caching didn't make it to Rx-version?

**Comment by sergey-visual-studio** [comment]
Total # of file in the solution and their language are the biggest variables. Once you get to 190 projects the # of projects themselves plays into that as well. Performance you describe in the latest version seems to be adequate enough w/o comparing it to the old version.

Converting from WinForms to WPF in DPack Rx does have a price to pay. Question would be whether it takes place during the solution processing/collection or later binding a large result set to WPF view. I'd suggest turning DPack Rx logging on, which can be found on Tools|Options -> DPack Rx General page. AppData\Local\DPack Rx is the location of log file. Collect the logs exercising File Browser and post it here. Thanks.

**Comment by frankfuu** [comment]
I also faced this issue. Initially I tried looking for ways to "Enable file structure caching" but couldnt' find it anywhere.

I ended up going back to Visual Studio 2019 because the faster (Alt+U) speeds were more useful to me compared to the new VS 2022 features.

**Comment by sergey-visual-studio** [comment]
> I also faced this issue. Initially I tried looking for ways to "Enable file structure caching" but couldnt' find it anywhere.
> 
> I ended up going back to Visual Studio 2019 because the faster (Alt+U) speeds were more useful to me compared to the new VS 2022 features.

Wow, that's pretty drastic. :) 

Let me take a look to see if it'd possible to move that to VS 2022 version. Not sure how quickly I'll be able to get to that though.

**Comment by frankfuu** [comment]
> Wow, that's pretty drastic. :)
> 
> Let me take a look to see if it'd possible to move that to VS 2022 version. Not sure how quickly I'll be able to get to that though.

Yeah DPack's Alt+U feature has saved me countless hours over the last decade or so. Would absoluately amazing if this feature can be ported to VS2022's version. Honestly happy to donate in any way or form.

**Comment by sergey-visual-studio** [comment]
> Yeah DPack's Alt+U feature has saved me countless hours over the last decade or so. Would absoluately amazing if this feature can be ported to VS2022's version.

Out of curiosity, what's your solution make up like? Meaning, projects language, number of projects, target framework and total number of files? Thanks.

**Comment by frankfuu** [comment]
> > Yeah DPack's Alt+U feature has saved me countless hours over the last decade or so. Would absoluately amazing if this feature can be ported to VS2022's version.
> 
> Out of curiosity, what's your solution make up like? Meaning, projects language, number of projects, target framework and total number of files? Thanks.

We target .net framework 48 using c# , have about 12 projects and the file breakdown with the following.

```
Name          Count
----          -----
.cs            2210
.cshtml        1665
.js            1057
.PNG            661
.jsx            263
.datasource     225
.scss           194
.css            167
.dll            123
.svg            104
.config         102
.less            98
.xml             93
.json            66
.html            54
.map             44
.gif             30
.csproj          29
.ttf             23
.txt             21
.sitemap         21
.woff            19
.ps1             15
.eot             15
.xsd             15
.woff2           12
.md              12
.otf              9
.ico              8
.pubxml           8
.svcinfo          6
.webmanifest      4
```

**Comment by sergey-visual-studio** [comment]
In my day job we have a few solutions, with the largest one being roughly 70 projects. All .NET 4.8 (unfortunately) as well. It packs slightly more code files than your test case, and large set of Xaml files. Yes, there is a slight delay opening it up but the total time is under 1 second, otherwise it would've shown the progress dialog.

So having said that, I wonder what makes your solution so different from my test case. What kind of performance were you getting on VS 2022 out of it?

**Comment by frankfuu** [comment]
> In my day job we have a few solutions, with the largest one being roughly 70 projects. All .NET 4.8 (unfortunately) as well. It packs slightly more code files than your test case, and large set of Xaml files. Yes, there is a slight delay opening it up but the total time is under 1 second, otherwise it would've shown the progress dialog.
> 
> So having said that, I wonder what makes your solution so different from my test case. What kind of performance were you getting on VS 2022 out of it?

I'll have to double check again when i'm on my work from office laptop during my office days, I'll report back in a few days ;)

**Comment by sergey-visual-studio** [comment]
Took a look at old source code and found what I was looking for. Doesn't look too complicated to port it to the new version. I'll just need to find some time to get to it.

**Comment by frankfuu** [comment]
> In my day job we have a few solutions, with the largest one being roughly 70 projects. All .NET 4.8 (unfortunately) as well. It packs slightly more code files than your test case, and large set of Xaml files. Yes, there is a slight delay opening it up but the total time is under 1 second, otherwise it would've shown the progress dialog.
> 
> So having said that, I wonder what makes your solution so different from my test case. What kind of performance were you getting on VS 2022 out of it?

On an older laptop it took about 2-2.5 seconds, on my beefier desktop it was under 1 second (but still noticable compared to VS2019 performance)

**Comment by sergey-visual-studio** [comment]
Guys, could you try the new beta posted in issue #54 ? It contains a different patch than what I'd originally planned here. It does prevent duplicate projects processing, which undoubtedly had contributed to this issue. Thanks.

**Comment by igitur** [comment]
I don't have large projects, so haven't really noticed this, but I'm just curious. Is the slowdown mentioned here due to the enumeration of the files, or the WPF overhead in populating the observable collections?

**Comment by sergey-visual-studio** [comment]
Yeah, large part of that doubling of processing time in some cases. WPF conversion did carry perf penalty.  There is another piece I started on but haven't gotten to completing is resurrecting File Browser project caching.

**Comment by igitur** [comment]
I see references to VirtualizingStackPanel in the ListView attributes, but I don't see a VirtualizingStackPanel anywhere. Is this correct?

**Comment by sergey-visual-studio** [comment]
It's created behind the scenes based on a couple of properties: VirtualizingStackPanel.IsVirtualizing and VirtualizingStackPanel.VirtualizationMode

**Comment by sergey-visual-studio** [comment]
Anybody interested in giving test version to try? Thanks.

**Comment by igitur** [comment]
> Anybody interested in giving test version to try? Thanks.

Sure, I'll test. Where do I find it?

**Comment by sergey-visual-studio** [comment]
Will post it this weekend.

**Comment by sergey-visual-studio** [comment]
Here you go:

[DPackBeta.zip](https://github.com/sergey-visual-studio/dpack/files/13323448/DPackBeta.zip)

**Comment by sergey-visual-studio** [comment]
Have you had a chance to try the new beta? Looking for some feedback.

**Comment by frankfuu** [comment]
> Here you go:
> 
> [DPackBeta.zip](https://github.com/sergey-visual-studio/dpack/files/13323448/DPackBeta.zip)

I can confirm this package feels a lot smoother for my use cases! Thanks so much for your efforts @sergey-visual-studio! 

Will you be publishing this to the official channels?

**Comment by danielHalan** [comment]
Can also confirm its working smoother...

**Comment by igitur** [comment]
Sorry for the late response. I never really had an issue with performance, because my solutions aren't very big. My issues were with duplicates. That said, I tested the beta anyway and it solves my duplicates problem and I don't see a change in performance, so I'm happy with it.

**Comment by sergey-visual-studio** [comment]
This should be in the newly posted release version.
